src/algokit_utils/_ensure_funded.py

```diff
@@ -49,52 +45,100 @@ class EnsureBalanceParameters:
 
-    :return None | PaymentTxn: None if balance was sufficient or the payment transaction used to increase the balance
-    """
-    address_to_fund = (
-        parameters.account_to_fund
-        if isinstance(parameters.account_to_fund, str)
-        else address_from_private_key(parameters.account_to_fund.private_key)  # type: ignore[no-untyped-call]
-    )
 
-    account_info = client.account_info(address_to_fund)
-    assert isinstance(account_info, dict)
+    Args:
+        client (AlgodClient): An instance of the AlgodClient class from the AlgoSDK library.
+        parameters (EnsureBalanceParameters): An instance of the EnsureBalanceParameters class that
+        specifies the account to fund and the minimum spending balance.
+
+    Returns:
+        PaymentTxn | str | None: If funds are needed, the function returns a payment transaction or a
+        string indicating that the dispenser API was used. If no funds are needed, the function returns None.
+    """
 
+    address_to_fund = _get_address_to_fund(parameters)
+    account_info = _get_account_info(client, address_to_fund)
     balance_micro_algos = account_info.get("amount", 0)
-    minimum_balance_micro_algos = account_info.get("min-balance")
+    minimum_balance_micro_algos = account_info.get("min-balance", 0)
     current_spending_balance_micro_algos = balance_micro_algos - minimum_balance_micro_algos
-    if parameters.min_spending_balance_micro_algos > current_spending_balance_micro_algos:
-        funding_source = parameters.funding_source or get_dispenser_account(client)
-        sender_address = address_from_private_key(funding_source.private_key)  # type: ignore[no-untyped-call]
+    fund_amount_micro_algos = _calculate_fund_amount(parameters, current_spending_balance_micro_algos)
 
-        min_fund_amount_micro_algos = parameters.min_spending_balance_micro_algos - current_spending_balance_micro_algos
-        fund_amount_micro_algos = max(min_fund_amount_micro_algos, parameters.min_funding_increment_micro_algos)
-        logger.info(
-            f"Funding {address_to_fund} {fund_amount_micro_algos}µ from {sender_address} to reach "
-            f"minimum spend amount of {parameters.min_spending_balance_micro_algos}µ (balance = "
-            f"{balance_micro_algos}µ, requirement = {minimum_balance_micro_algos}µ)"
-        )
-        return transfer(
-            client,
-            TransferParameters(
-                from_account=funding_source,
-                to_address=address_to_fund,
-                micro_algos=fund_amount_micro_algos,
-                note=parameters.note or "Funding account to meet minimum requirement",
-                suggested_params=parameters.suggested_params,
-                max_fee_micro_algos=parameters.max_fee_micro_algos,
-                fee_micro_algos=parameters.fee_micro_algos,
-            ),
-        )
+    if fund_amount_micro_algos is not None:
+        if is_testnet(client) and parameters.use_dispenser_api:
```

Comment:
We potentially don't even need `parameters.use_dispenser_api`. Instead we could check for the presence of `ALGOKIT_DISPENSER_ACCESS_TOKEN` and fund using the dispenser API

Reply:
I think it's worthwhile opting in to using the dispenser, I had a suggestion in the typescript one for a clean way of doing it using fundingSource parameter.

Reply:
@robdmoore @neilcampbell instead of just str i am actually setting a class of dispenser client. 

So to user, user just instantiates the dispenser client (where constructor will auto load CI value for token OR you can pass value directly to constructor). Then he sets that client instance as a funding source, within the ensure funded that instance is used to call the fund method. 

I think that was a better use of a client considering we need to introduce a api client class anyway.

Let me know what you think

Reply:
s/he/they/

but yep, sounds like a good idea, nice